simplicity-sys/src/ffi.rs

```diff
@@ -0,0 +1,480 @@
+        pub case_cnt: size_t,
+        pub pair_cnt: size_t,
+        pub disconnect_cnt: size_t,
+        pub injl_cnt: size_t,
+        pub injr_cnt: size_t,
+        pub take_cnt: size_t,
+        pub drop_cnt: size_t,
+    }
+
+    /// A source and target type (in the C code this is actually an anonymous struct
+    /// where it is used).
+    #[repr(C)]
+    #[derive(Copy, Clone, Eq, PartialEq, Debug)]
+    pub struct CArrow {
+        pub source_type: size_t,
+        pub target_type: size_t,
+    }
+
+    #[repr(C)]
+    #[derive(Copy, Clone)]
+    /// Anonymous
+    pub union CAuxTypes {
+        /// scratch space for verifyCanonicalOrder
+        pub aux: size_t,
+        /// source type, target type
+        pub types: [size_t; 2],
+    }
+
+    #[repr(C)]
+    #[derive(Copy, Clone, Eq, PartialEq, Debug)]
+    /// Anonymous
+    pub struct CSourceIx {
+        source_ix: size_t,
+    }
+
+    #[repr(C)]
+    #[derive(Copy, Clone, Eq, PartialEq, Debug)]
+    /// Anonymous
+    pub struct CChild2 {
+        child: [size_t; 2],
+    }
+
+    #[repr(C)]
+    #[derive(Copy, Clone)]
+    /// Anonymous
+    pub union CSourceChildValue {
+        source_ix: CSourceIx,
+        child: CChild2,
+        compact_value: CBitstring,
+    }
+
+    #[repr(C)]
+    #[derive(Copy, Clone)]
+    pub struct CDagNode {
+        pub jet: *const c_void,
+        pub cmr: CSha256Midstate,
+        pub aux_types: CAuxTypes,
+        pub source_ix_or_child_or_compact_value: CSourceChildValue,
+        pub target_ix: size_t,
+        pub cost: super::ubounded,
+        pub tag: CTag,
+    }
+
+    #[repr(C)]
+    #[derive(Copy, Clone, Eq, PartialEq, Debug, Default)]
+    /// Static analyses for a particular node of a Simplicity DAG
+    pub struct CAnalyses {
+        pub annotated_merkle_root: CSha256Midstate,
+    }
+
+    extern "C" {
+        /// Given the IMR of a jet specification, return the CMR of a jet that implements
+        /// that specification
+        pub fn mkJetCMR(imr: *const u32) -> CSha256Midstate;
+
+        /// Compute the CMR of a jet of scribe(v) : ONE |- TWO^(2^n) that outputs a given
+        /// bitstring
+        pub fn computeWordCMR(value: *const CBitstring, n: size_t) -> CSha256Midstate;
+
+        /// Given a well-formed dag[i + 1], set the `cmr` field of every node in `dag`
+        pub fn computeCommitmentMerkleRoot(dag: *mut CDagNode, i: size_t);
+
+        /// Given a well-typed dag representing a Simplicity expression, compute
+        /// the annotated Merkle roots of all subexpressions.
+        pub fn computeAnnotatedMerkleRoot(
+            analyses: *mut CAnalyses,
+            dag: *const CDagNode,
+            ty: *const CType,
+            len: size_t,
+        );
+
+        /// Verifies that the 'dag' is in canonical order, meaning that nodes
+        /// under the left branches have lower indices than nodes under
+        pub fn verifyCanonicalOrder(dag: *const CDagNode, len: size_t) -> SimplicityErr;
```

Comment:
The C comment says that `dag` here might change in some cases and does not mark it as const pointer. We should also make this mut *